CommandDotNet/MicrosoftCommandLineUtils/CommandLineApplication.cs

```diff
@@ -154,101 +148,91 @@ public int Execute(params string[] args)
             for (var index = 0; index < args.Length; index++)
             {
                 var arg = args[index];
-                var processed = false;
-                if (!processed && option == null)
+                
+                /* Process flow
+                 *
+                 * easy to determine options so check those first.  options start with -- or -
+                 * if option,
+                 *   check if the value is provided in same string with : or =
+                 *     if value is needed and not provided store option for next argument
+                 *
+                 * if previous argument was option needing a value,
+                 *   parse argument for option
+                 *
+                 * if argument is the name of a subcommand
+                 *   processing remaining arguments for the sub command
+                 *
+                 * assign argument as value for next argument in the command
+                 */
+                
+                // control flow is tricky.  Watch for: continue, break, return & throw
+                
+                if (option == null)
                 {
-                    string[] longOption = null;
-                    string[] shortOption = null;
+                    string[] optionParts = null;
 
-                    if (arg.StartsWith("--"))
+                    var isLongOption = arg.StartsWith("--");
+                    var isShortOption = !isLongOption && arg.StartsWith("-");
+                    
+                    if (isLongOption)
                     {
-                        longOption = arg.Substring(2).Split(new[] { ':', '=' }, 2);
-                    }
-                    else if (arg.StartsWith("-"))
-                    {
-                        shortOption = arg.Substring(1).Split(new[] { ':', '=' }, 2);
-                    }
-                    if (longOption != null)
-                    {
-                        processed = true;
-                        var longOptionName = longOption[0];
-                        option = command.GetOptions().SingleOrDefault(opt => string.Equals(opt.LongName, longOptionName, StringComparison.Ordinal));
+                        optionParts = arg.Substring(2).Split(new[] { ':', '=' }, 2);
+                        var optionName = optionParts[0];
+                        option = command.FindOption(o => o.LongName, optionName);
 
                         if (option == null)
                         {
-                            if (string.IsNullOrEmpty(longOptionName) && !_appSettings.ThrowOnUnexpectedArgument  && _appSettings.AllowArgumentSeparator)
+                            var isArgumentSeparator = string.IsNullOrEmpty(optionName);
+                            
+                            // ??? if AllowArgumentSeparator is true, why would this be considered an unexpected argument?
```

Comment:
@bilal-fazlani if AllowArgumentSeparator is true, why would this be considered an unexpected argument?

Reply:
hmmm everything in the directory MicrosoftCommandLineUtils has been a black box for me. I copied this from a microsoft nuget package.

I will have to analyse this code to understand more.

Reply:
This makes more sense to me, what do you think?

```c#
if (isArgumentSeparator && _appSettings.AllowArgumentSeparator)
{
    // skip over the '--' argument separator
    // all remaining arguments to be added to command.RemainingArguments 
    index++;
}

HandleUnexpectedArg(command, args, index, argTypeName: "option");
break;
```

Reply:
Yes makes sense.